crates/librqbit/src/torrent_state/live/mod.rs

```diff
@@ -795,6 +795,86 @@ impl TorrentStateLive {
             .take_while(|r| r.is_ok())
             .last();
     }
+
+    async fn task_send_pex_to_peer(
+        self: Arc<Self>,
+        peer_addr: SocketAddr,
+        tx: PeerTx,
+    ) -> anyhow::Result<()> {
+        let mut sent_peers_live: HashSet<SocketAddr> = HashSet::new();
+        const MAX_SENT_PEERS: usize = 50; // As per BEP 11 we should not send more than 50 peers at once (here it also applies to fist message, should be OK as we anyhow really have more)
+        const PEX_MESSAGE_INTERVAL: Duration = Duration::from_secs(60); // As per BEP 11 recommended interval is min 60 seconds
+        let mut delay = Duration::from_secs(10); // Wait 10 seconds before sending the first message to assure that peer will stay with us
+
+        loop {
+            tokio::select! {
+                _ = tx.closed() => return Ok(()),
+                _ = tokio::time::sleep(delay) => {},
+
+            }
+            delay = PEX_MESSAGE_INTERVAL;
+
+            let addrs_live_to_sent = self
+                .peers
+                .states
+                .iter()
+                .filter_map(|e| {
+                    let peer = e.value();
+                    let addr = peer.outgoing_address.as_ref().unwrap_or_else(|| e.key());
+
+                    if *addr != peer_addr {
+                        let has_outgoing_connections = peer
+                            .stats
+                            .counters
+                            .outgoing_connections
+                            .load(Ordering::Relaxed)
+                            > 0; // As per BEP 11 share only those we were able to connect
+                        if peer.state.is_live()
+                            && has_outgoing_connections
+                            && !sent_peers_live.contains(addr)
+                        {
+                            Some(*addr)
+                        } else {
+                            None
+                        }
+                    } else {
+                        None
+                    }
+                })
+                .take(50)
+                .collect::<HashSet<_>>();
+
+            let addrs_closed_to_sent = sent_peers_live
+                .iter()
+                .filter(|addr| {
+                    self.peers
+                        .states
+                        .get(addr)
```

Comment:
you're locking the state twice, and it may race - e.g. it'll be live in first pass where you are computing "addrs_live_to_sent" and then not live later, or the other way around.

It's hard to reason about what will happen in this case, esp. when below you're trying to provide strong guarantees like "it's assured by mutual exclusion of two  above sets".

So overall, how about refactor this for simplicity:

1. store live peer's view in "sent_peers_live" as you are doing now.
1. for each loop iteration
    1. collect **all** live addrs into HashSet (not bounded by 50, but bounded by the semaphore anyway, 128 if I remember correctly). Let's name it "all_live"
    1. connected = (all - sent).take(50)
    2. dropped = (sent - all).take(50)
    3. sent = sent + connected - dropped

Reply:
I see - but I think it's not the case here because liveness check actually happens only once, depending on content of `sent_peers_live` - which reflects already sent peers addresses:
- if address is in `sent_peers_live`, then its effectively checked for liveness only on line 853 `.map(|p| !p.value().state.is_live())`  (for closed peers), because of line 834 `&& !sent_peers_live.contains(addr)` which excludes it ultimately from `addrs_live_to_sent` set.
- if address is not in `sent_peers_live` -  it's checked only for `addrs_live_to_sent` on line 832 `if peer.state.is_live()` and second check for closed peers is not happening at all, because it's in iteration over `sent_peers_live`.

So in any case in each loop run **only one liveness check happens**, which is relevant for logic.

Reply:
My main point was that the logic is hard to read, and I proposed a simpler way to write it. All the state that you need to keep in your mind to understand why the liveness check is happening only once might serve as proof of that :)

Reply:
I do not think that logic is difficult -   it's just all about `sent_peers_live`, which we need anyhow - if address is not in it, we check if we can add new address, which is live and not in it,   if address is in it, we check if we should remove if it's not live. That's it.  

Proposed change will work exactly same,   I think. Only  we will need always to collect all live peers, which is not "needed", we just need those, that are not in `sent_peers_live`.